### ISSUE TITLE
[dask-on-ray] Zero-copy deserialization of dask.delayed() objects?

### ISSUE THREAD

**Issue by Hoeze** [ray-project/ray#13620]
Hi, I just tried using the dask-on-ray scheduler and I'm mind blown how well this works!
I had quite a number of issues with dask's distributed scheduler that are nearly completely solved by the shared worker memory and the zero-copy transfers.

However, I noticed that when I want to pass a big object to a function with dask.delayed(), it gets de-serialized on every worker and therefore results in `num_workers` copies of the dataset.
For example:
```python
big_obj = np.random.randint(10**9)
delayed_big_obj = dask.delayed(big_obj, pure=True)

@dask.delayed
def some_processing_fn(idx, big_obj=delayed_big_obj):
    return big_obj[idx]

tasks = [some_processing_fn(idx) for idx in range(1000)]

dask.compute(tasks, scheduler=ray_dask_get)
```

Instead, if I use `ray.put()` / `ray.get()`, the workers directly use the shared-memory object:
```python
big_obj = np.random.randint(10**9)
big_obj_id = ray.put(big_obj)

@dask.delayed
def some_processing_fn(idx, big_obj_id=big_obj_id):
    return ray.get(big_obj_id)[idx]

tasks = [some_processing_fn(idx) for idx in range(1000)]

dask.compute(tasks, scheduler=ray_dask_get)
```
This second approach made my script use 20x less memory, but I had to change everything from dask.delayed() to ray.put()/ray.get()...

**Comment by clarkzinzow** [comment]
Hi @Hoeze, I'm happy to hear that Dask-on-Ray is working pretty well for you!

**TLDR;** You should be able to get around this in a slightly nicer way by not using Dask-delayed or Ray objects as **default arguments** to other Dask-delayed functions, instead passing in Ray objects (returned from `ray.put()`) as explicit arguments to those Dask-delayed functions. The good news is you shouldn't have to do a `ray.get()` within `some_processing_fn` if `big_obj` is given as an explicit argument, _not_ a default argument.

There are a few unfortunate things going on here! By giving `delayed_big_obj` as a default argument to `some_processing_fn`, `delayed_big_obj` is considered to be part of the function _definition_. Unfortunately, given that Dask-on-Ray proxies all Dask tasks through a single `task_executor` remote function by passing the pickled Dask task function and arguments to `task_executor.remote()`, all in an effort to cut down on function exporting overhead,`delayed_big_obj` being part of the function definition means that it gets serialized as part of the function definition, _not_ as a function argument, and therefore we're deserializing a large NumPy array from the function definition on every invocation of `some_processing_fn`.

The reason that `ray.put(big_obj)` bypasses this issue is because it is only the Ray object ID that's serialized, not the actual object value, which is stored in shared memory via the put. However, because it's a default argument to the Dask-delayed function instead of an _explicit_ argument, Dask-on-Ray doesn't automatically deserialize it, making the `ray.get()` in `some_processing_fn` necessary. This is definitely something that Dask-on-Ray should do for you.

You should note that, even if you give `delayed_big_obj` as an explicit argument to `some_processing_fn`, it will end up making a **separate** shared memory object for each invocation of `some_processing_fn`. Why? Because Dask-on-Ray doesn't realize that `some_processing_fn(idx, delayed_big_obj)` is being passed the same `delayed_big_obj` for every invocation, and doesn't create a single shared memory object for all invocations. I also think that this is something that Dask-on-Ray should do for you, since it's a large performance and resource utilization improvement for workloads like yours that involve large fan-out of source data.

### Workaround

You should be able to work around both the default argument issue and the shared memory object duplicate issue by passing in Ray objects (returned from `ray.put()`) as explicit arguments to those Dask-delayed functions, and you shouldn't have to do a `ray.get()` within `some_processing_fn` if the Ray object is given as an explicit argument. E.g., the following should work:
```python
big_obj = np.random.randint(10**9)
big_obj_id = ray.put(big_obj)

@dask.delayed
def some_processing_fn(idx, big_obj):
    return big_obj[idx]

tasks = [some_processing_fn(idx, big_obj_id) for idx in range(1000)]
dask.compute(tasks, scheduler=ray_dask_get)
```
Please let me know if that works for you. If it doesn't, that is probably another thing in Dask-on-Ray that needs to be fixed.

### Long-term Solution

1. Dask-on-Ray should automatically link in default arguments as plain task arguments at Ray task submission time.
2. Dask-on-Ray should transparently do that `ray.put()` for you for large source objects that are part of a large fan-out.

**Comment by Hoeze** [comment]
Thanks for the very concise explanation @clarkzinzow!
Yes, I can confirm that your solution works :)
(Note that it should be `np.random.randint(0, 100, 10**9)` though, sorry for the untested code ^^)
